posts/dumb_things_with_postgres.md

```diff
@@ -0,0 +1,77 @@
+
+```ruby
+class Article < ActiveRecord::Base
+  # all the other logic
+
+  def associated_articles
+    Article.where(id != self.id).where("tags = ANY(#{self.tags})")
+  end
+end
+```
+
+If you've been using Rails/Postgres over the past couple of years this might be very run of the mill so lets dive into some of the madness you can cause.
+
+Example:
+
+A customer already has a series of one to one relations like Products to Product Categories but they now want this to be a many to many relation. They've already input lots of data and if possible they want to keep this. The obvious solution would be to create a joining table, iterate over the existing products and create entries for each of the associations and then drop the columns, I instead decided to do something a bit more out there, what if we could convert this integer column over to an array column while keeping all the existing data and not have to iterate over everything that's already there. To do this you'll need a really weird migration like this;
+
+```ruby
+class ChangeXIdColumn < ActiveRecord::Migration
+  def up
+    remove_index :products, :product_category_id
+    change_column :products, :product_category_id, 'integer[] USING ARRAY[product_category_id]::INTEGER[]', array: true, null: false, default: []
+    rename_column :products, :product_category_id, :product_category_ids
+    add_index :products, :product_category_ids, using: 'gin'
+  end
+
+  def down
+    raise ActiveRecord::IrreversibleMigration
+  end
+end
+```
+
+There's a lot going on in this migration so I'll break it down a bit.
+```ruby
+  change_column :products, :product_category_id, 'integer[] USING ARRAY[product_category_id]::INTEGER[]', array: true, null: false, default: []
+```
+This line takes our existing foreign key column and converts it to and array type, the SQL being executed recasts the existing data in the correct format so `4` would become `{4}`. The curly braces are how Postgres denotes an array, when returned by ActiveRecord the data would be represented as a standard array.
+
+```ruby
+  add_index :products, :product_category_ids, using: 'gin'
+```
+This is clearly an index, the important part however is that it's using "GIN" as it's index type. GIN is lossless versus GiST, the other index type available, which is lossy introducting the chance for incorrect results slightly. In practice it probably won't make a huge amount of difference to you which one you use but if you want to read more on these types I'd recommend the [Postgres documentation](https://www.postgresql.org/docs/current/static/textsearch-indexes.html). One thing to watch out for is that GIN indexes can have long build times but there are ways to mitigate that.
```

Comment:
I'd mention it is a full-text index type somewhere, I didn't realise until following the linkaroo.